### DISCUSSION TITLE
PowerShell script, and a bit more lol

### DISCUSSION THREAD

**Discussion by vt-idiot** [s0hv/foo_discord_rich#6] [Show and tell]
```PS
$upload = Resolve-Path -LiteralPath $input | Select-Object -ExpandProperty Path
$uri = "https://catbox.moe/user/api.php"
$body = @{
    "reqtype"="fileupload"
    "fileToUpload"=Get-Item $upload
}
Invoke-WebRequest -Uri $uri -Method POST -Form $body | Select-Object -ExpandProperty Content
```

Works great. No API keys needed, no fuss. Intentionally omitted a userhash, because if someone else has uploaded the same file (without being logged in), catbox API will just return that pre-existing URL.

I have the script saved as "catbox.ps1" so I just set the Artwork upload command to: `pwsh.exe C:\catbox.ps1`

If you aren't using PowerShell 7 it won't work. Couldn't figure out why. Also, you will need to open PowerShell first and either `Set-ExecutionPolicy RemoteSigned -Scope CurrentUser` or run it as an Administrator and `Set-ExecutionPolicy RemoteSigned` but you already knew that if you wanted to use a PowerShell script.

---

Does Discord actually cache the images? If so, it might make sense to use litterbox. litterbox allows temporary anonymous uploads. If 1hr is too short, there are other allowed values. from https://litterbox.catbox.moe/tools.php:
> ## The API is located at https://litterbox.catbox.moe/resources/internals/api.php
> There is only 1 request type for Litterbox - fileupload. All uploads are anonymous (you don't need an account/userhash).
> ## File uploads
> ### File Uploads
> `reqtype="fileupload" fileToUpload=(file data here)`
> ## cURL to API
> If you want to make curl requests to the API, here is an example. Allowed values for "time" are 1h, 12h, 24h, and 72h.
> `curl -F "reqtype=fileupload" -F "time=1h" -F "fileToUpload=@cutie.png" https://litterbox.catbox.moe/resources/internals/api.php`

curl is already in my PATH but I'm honestly not 100% sure how to pass stdin input into the form line. Faced with writing a .bat script for something as silly as selectively parsing stdin within a quoted string, I decided it was easier* to just use PowerShell...

```PS
$upload = Resolve-Path -LiteralPath $input | Select-Object -ExpandProperty Path
$uri = "https://litterbox.catbox.moe/resources/internals/api.php"
$body = @{
	"reqtype"="fileupload"
	"time"="1h"
	"fileToUpload"=Get-Item -Path $upload
}
Invoke-WebRequest -Uri $uri -Method POST -Form $body | Select-Object -ExpandProperty Content
```

**Comment by s0hv** [comment]
This is pretty nice. I tested it with powershell 7.3 and it mostly works. Might work with powershell 6 as well but earlier versions than that don't have the `-Form` parameter. The script worked as long as the path to the cover art file does not contain UTF-8 characters like kanji. If it contains those kind of characters then the upload will fail. However, the process will still exit with code 0 making the plugin think the process exited successfully with a url (This part could use some more checks for a valid url in the plugin itself so exit code would matter less.). Using cURL has the exact same problem with UTF-8 paths and the last time I checked it's not really possible to fix that with cURL. That's the main reason why my initial idea was to use Python as it would make handling UTF-8 much much easier.

As for limited time uploads. Discord serves the images through their CDN so I would imagine some caching being in place. However, the plugin saves the urls and assumes they are permanent. It also caches the `image md5 hash - url` pair so if the same image is uploaded again it will use the existing url. This makes using limited time urls not possible within the current feature set.

**Reply by vt-idiot** [reply]
> The script worked as long as the path to the cover art file does not contain UTF-8 characters like kanji.

It only failed for me when the path contained Japanese UTF-8 characters that were stand-ins for characters that "aren't allowed" in paths. ／　and　：。**Everything** I tested it on had kanji, kana, or both in the path lol.

I long looooong ago changed my system codepage to 65001 and AFAIK PS7 handles UTF-8 with ease.

**Reply by vt-idiot** [reply]
Thanks for clarifying on the limited time URLs! Manually edited the JSON hash list to remove my litterbox links from testing and just switched to using catbox proper for now.

**Reply by s0hv** [reply]
It might be an encoding problem, as the script worked fine when I echoed the path to the script but failed when the plugin did it.

**Reply by s0hv** [reply]
```
 Resolve-Path: C:\Users\s0hv\Downloads\???\catbox.ps1:1
Line |
   1 |  $upload = Resolve-Path -LiteralPath $input | Select-Object -ExpandPro .
     |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot find path 'C:\Users\s0hv\Music\Test\Junko Ohashi (大橋純子) test\cover.jpg' because it does not
     | exist.
Get-Item: C:\Users\s0hv\Downloads\???\catbox.ps1:5
Line |
   5 |      "fileToUpload"=Get-Item $upload
     |                              ~~~~~~~
     | Cannot bind argument to parameter 'Path' because it is null.
``` 
This is the error message I get btw. And the path most certainly does exist. If I copy paste it into explorer it will open the file just fine. Also tried moving the script out of the folder with special characters and that didn't help either. Weird issue.

**Reply by vt-idiot** [reply]
When I sent paths to the script manually while testing, I did it with files that were right next to the script. All the Unicode paths I tried were piped directly from the plugin.

![image](https://user-images.githubusercontent.com/81622808/218301026-debc959d-cea7-407d-97f7-f9f77f31db59.png)

That's what I mean about my system-wide codepage being UTF-8.

![image](https://user-images.githubusercontent.com/81622808/218301377-51e29735-f7ae-49c7-ab92-865f9e6acb58.png)

First time seeing that file already in use error. Switching tracks within the same album made it try again and then it worked...

![image](https://user-images.githubusercontent.com/81622808/218301330-36aaaabb-7f58-4bee-9b1f-e6e27fe826c8.png)

**Reply by vt-idiot** [reply]
> ```
>  Resolve-Path: C:\Users\s0hv\Downloads\???\catbox.ps1:1
> Line |
>    1 |  $upload = Resolve-Path -LiteralPath $input | Select-Object -ExpandPro .
>      |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>      | Cannot find path 'C:\Users\s0hv\Music\Test\Junko Ohashi (大橋純子) test\cover.jpg' because it does not
>      | exist.
> Get-Item: C:\Users\s0hv\Downloads\???\catbox.ps1:5
> Line |
>    5 |      "fileToUpload"=Get-Item $upload
>      |                              ~~~~~~~
>      | Cannot bind argument to parameter 'Path' because it is null.
> ```
> 
> This is the error message I get btw. And the path most certainly does exist. If I copy paste it into explorer it will open the file just fine. Also tried moving the script out of the folder with special characters and that didn't help either. Weird issue.

~~Just saw that while I was making my screenshot. I think it might be because of the (parenthesis), not because of the kanji/kana.~~

Nope. Works.

![image](https://user-images.githubusercontent.com/81622808/218301641-9a531545-2438-463d-a33f-f5bce8c20f68.png)

**Reply by s0hv** [reply]
That might be true. I'll test it more later today if I remember.

**Reply by vt-idiot** [reply]
Don't get me wrong btw, I don't think anyone writes a PowerShell script to do something _by choice_

I know how to IWR a POST request, and I already had a separate (set) of scripts I use with catbox + my userhash, so that part was already done. No idea how to upload a file using a Python script, else I'd have just re-written the imgur one to do the same thing lol.

**Reply by vt-idiot** [reply]
While I'm here.

```
Opening track for playback: "O:\OneDrive\Music\宇多田ヒカル\DEEP RIVER (2018 Remastered Album)\13. Simple And Clean.flac"
Discord Rich Presence Integration v1.3.0-beta+f6c78e4: error 4000
child "activity" fails because [child "state" fails because ["state" length must be at least 2 characters long]]
```

Not sure what that one is about. It already has a URL showing in the File Info and I can see it in the JSON too.

**Reply by vt-idiot** [reply]
Oh. I think I know why, lol
```
child "activity" fails because [child "state" fails because ["state" length must be at least 2 characters long]]
Opening track for playback: "O:\OneDrive\Music\宇多田ヒカル\DEEP RIVER (2018 Remastered Album)\12. 光.flac"
```
"Hikari" is one character, `光`

...there's probably a way I can pad the foobar %title% string directly from the plugin settings.

**Reply by vt-idiot** [reply]
`[$pad(%title%,2)]` - should probably be the default setting...
![image](https://user-images.githubusercontent.com/81622808/218302239-cff0e6ca-524e-4d31-ab70-430894f1abeb.png)

**Reply by s0hv** [reply]
> When I sent paths to the script manually while testing, I did it with files that were right next to the script. All the Unicode paths I tried were piped directly from the plugin.
> 
> ![image](https://user-images.githubusercontent.com/81622808/218301026-debc959d-cea7-407d-97f7-f9f77f31db59.png)
> 
> That's what I mean about my system-wide codepage being UTF-8.

After settings this option on everything started working. Seems like UTF-8 string handling differs when it's done between two different processes compared to doing with echo.

If it's okay I'll add the script to the [documentation](https://s0hv.github.io/foo_discord_rich/docs/album-artwork/#upload-scripts) in case someone wants to use powershell for this. Having multiple different options for this is always better than less. If you would create a gist of that script that'd probably be best so I can just link to that in the documentation and the author of it would be clear.

**Reply by s0hv** [reply]
> Oh. I think I know why, lol
> 
> ```
> child "activity" fails because [child "state" fails because ["state" length must be at least 2 characters long]]
> Opening track for playback: "O:\OneDrive\Music\宇多田ヒカル\DEEP RIVER (2018 Remastered Album)\12. 光.flac"
> ```
> 
> "Hikari" is one character, `光`
> 
> ...there's probably a way I can pad the foobar %title% string directly from the plugin settings.

I had missed this edge case, good find. I'll see if pad is the best way to fix that.

**Reply by vt-idiot** [reply]
> After settings this option on everything started working. Seems like UTF-8 string handling differs when it's done between two different processes compared to doing with echo.

Yeah, Windows still has baggage from "using UTF-16LE for some things" and "inexplicably using ANSI/OEM codepages for other things". System wide UTF-8 fixes a lot of issues like this, but can break some legacy applications.
 
> If it's okay I'll add the script to the [documentation](https://s0hv.github.io/foo_discord_rich/docs/album-artwork/#upload-scripts) in case someone wants to use powershell for this. Having multiple different options for this is always better than less. If you would create a gist of that script that'd probably be best so I can just link to that in the documentation and the author of it would be clear.

Of course, [dozo](https://gist.github.com/vt-idiot/8a7161a48dc6f7f7719423e938217267)

**Reply by s0hv** [reply]
> ```
> child "activity" fails because [child "state" fails because ["state" length must be at least 2 characters long]]
> Opening track for playback: "O:\OneDrive\Music\宇多田ヒカル\DEEP RIVER (2018 Remastered Album)\12. 光.flac"
> ```

I fixed this error in the latest release. The padding step was already done in the code but it did not work correctly for utf-8 characters longer than a byte.
